File: src/api/cluster.rs
//! Running a server across several worker threads.
//!
//! [`Cluster`] is what [`Server::run`] hands back: the worker threads, their
//! shutdown switch, and the addresses they bound. [`cores`] is the worker
//! count to reach for when there is no better number.
//!
//! [`Server::run`]: crate::api::server::Server::run

/// How many threads the machine can run at once, or 1 if that cannot be found.
///
/// Useful as the worker count for [`Server::run`].
///
/// [`Server::run`]: crate::api::server::Server::run
pub fn cores() -> usize {
    std::thread::available_parallelism().map(|count| count.get()).unwrap_or(1)
}

/// A server running across several threads, as [`Server::run`]
/// returns it.
///
/// Dropping this leaves the workers running; call [`Cluster::close`] to wind
/// them down.
///
/// [`Server::run`]: crate::api::server::Server::run
pub struct Cluster {
    shutdown: tokio::sync::watch::Sender<Option<f64>>,
    threads: Vec<std::thread::JoinHandle<()>>,
    addresses: Vec<std::net::SocketAddr>,
}

impl Cluster {
    /// A cluster over already-started workers.
    pub fn new(shutdown: tokio::sync::watch::Sender<Option<f64>>, threads: Vec<std::thread::JoinHandle<()>>, addresses: Vec<std::net::SocketAddr>) -> Self {
        Self { shutdown, threads, addresses }
    }

    /// The first bound address, if any port has one.
    pub fn address(&self) -> Option<std::net::SocketAddr> {
        self.addresses.first().copied()
    }

    /// Every bound address.
    pub fn addresses(&self) -> &[std::net::SocketAddr] {
        &self.addresses
    }

    /// How many worker threads are running.
    pub fn workers(&self) -> usize {
        self.threads.len()
    }

    /// Stops every worker and waits for the threads to finish.
    ///
    /// `timeout` is passed to each worker's [`ServerHandle::close`], bounding
    /// how long it waits for its connections. This blocks, so do not call it
    /// from inside an async context.
    ///
    /// [`ServerHandle::close`]: crate::api::server::ServerHandle::close
    pub fn close(self, timeout: Option<f64>) {
        let _ = self.shutdown.send(timeout);
        for thread in self.threads {
            let _ = thread.join();
        }
    }
}

File: python/src/soyokaze/api/cluster.py
"""Running a server across several worker threads.

:class:`Cluster` is what :meth:`Server.run <soyokaze.api.server.Server.run>`
hands back: the worker threads and the ports they bound.
:meth:`Cluster.cores` is the worker count to reach for when there is no
better number.
"""

from ..ffi import library

class Cluster:
    """A server running across several threads, as :meth:`Server.run` returns it.

    :meth:`Server.run`: soyokaze.api.server.Server.run
    """

    def __init__(self, handle, callbacks):
        self.handle = handle
        self.callbacks = callbacks

    @classmethod
    def cores(cls):
        """How many threads the machine can run at once, or 1 if that cannot be found."""
        return library.soyokaze_cores()

    @property
    def port(self):
        """The port the first listener actually bound."""
        return library.soyokaze_cluster_port(self.handle)

    def ports(self):
        """The port of every bound address."""
        count = library.soyokaze_cluster_address_count(self.handle)
        return [library.soyokaze_cluster_port_at(self.handle, index) for index in range(count)]

    def address(self):
        """The address the first listener bound, or ``None``."""
        addresses = self.addresses()
        return addresses[0] if addresses else None

    def addresses(self):
        """Every address the cluster bound, as text."""
        count = library.soyokaze_cluster_address_count(self.handle)
        return [library.soyokaze_cluster_address_at(self.handle, index).take().decode() for index in range(count)]

    def workers(self):
        """How many worker threads are running."""
        return library.soyokaze_cluster_workers(self.handle)

    def close(self, timeout=None):
        """Stops every worker and waits for the threads to finish. This blocks."""
        if self.handle:
            library.soyokaze_cluster_close(self.handle, -1.0 if timeout is None else timeout)
            self.handle = None
